File: backtest/mce_sweep.cpp
// mce_sweep.cpp -- MCE threshold sweep in C++
// Finds optimal ATR/drift/vol thresholds for MacroCrashEngine.
// Mirrors MCE logic exactly: EWM drift 30s halflife, vol ratio 5s/60s,
// ATR M1 EMA-20. Entry on expansion + threshold breach. Trail SL.
//
// Build: g++ -O3 -std=c++17 mce_sweep.cpp -o mce_sweep
// Run:   ./mce_sweep ~/Tick/2yr_XAUUSD_tick.csv

#include <cstdio>
#include <cstdlib>
#include <cstring>
#include <cmath>
#include <cstdint>
#include <vector>
#include <array>
#include <algorithm>

struct Tick { int64_t ts_ms; double bid, ask; };

static int64_t parse_ts(const char* d, const char* t) {
    auto gi=[](const char* s,int n){int v=0;for(int i=0;i<n;i++)v=v*10+(s[i]-'0');return v;};
    int y=gi(d,4),mo=gi(d+4,2),dy=gi(d+6,2),h=gi(t,2),mi=gi(t+3,2),se=gi(t+6,2);
    if(mo<=2){y--;mo+=12;}
    int64_t days=365LL*y+y/4-y/100+y/400+(153*mo+8)/5+dy-719469LL;
    return(days*86400LL+h*3600LL+mi*60LL+se)*1000LL;
}

static std::vector<Tick> load(const char* path) {
    FILE* f=fopen(path,"r");if(!f){printf("Cannot open %s\n",path);exit(1);}
    std::vector<Tick> v;v.reserve(120000000);
    char line[256],d[32],t[32];double bid,ask;
    fgets(line,256,f);
    while(fgets(line,256,f)){
        if(sscanf(line,"%[^,],%[^,],%lf,%lf",d,t,&bid,&ask)<4)continue;
        if(bid<=0||ask<=bid)continue;
        v.push_back({parse_ts(d,t),bid,ask});
    }
    fclose(f);return v;
}

// Time-based EWM alpha
static inline double tba(double dt_ms, double hl_ms) {
    return 1.0 - exp(-dt_ms * 0.693147 / hl_ms);
}

struct Indicators {
    double ewm_mid=0, ewm_drift=0;
    double vs=0, vl=0, vol_ratio=1.0;
    double atr=0;
    double bo=0,bh=0,bl=0,bc=0; int64_t bmin=-1; bool bhas=false;
    int64_t prev_ms=0;
    static constexpr int64_t GAP_MS=3600000LL;

    void update(double bid, double ask, int64_t ts) {
        const double mid=(bid+ask)*0.5;
        if(prev_ms>0 && (ts-prev_ms)>GAP_MS) {
            ewm_mid=mid; ewm_drift=0; vs=vl=0; vol_ratio=1;
        }
        const double dt=prev_ms>0?(double)(ts-prev_ms):100.0;
        prev_ms=ts;

        // EWM drift 30s halflife
        const double a30=tba(dt,30000.0);
        ewm_mid=a30*mid+(1-a30)*ewm_mid;
        ewm_drift=mid-ewm_mid;

        // Vol ratio 5s/60s
        const double am=fabs(ewm_drift);
        const double a5=tba(dt,5000.0), a60=tba(dt,60000.0);
        vs=a5*am+(1-a5)*vs; vl=a60*am+(1-a60)*vl;
        vol_ratio=vl>1e-9?vs/vl:1.0;

        // ATR M1 EMA-20
        const int64_t bm=ts/60000LL;
        if(!bhas){bo=bh=bl=bc=mid;bmin=bm;bhas=true;}
        else if(bm!=bmin){
            const double rng=bh-bl;
            if(atr==0)atr=rng; else{const double a=2.0/21.0;atr=a*rng+(1-a)*atr;}
            bo=bh=bl=bc=mid;bmin=bm;
        } else{if(mid>bh)bh=mid;if(mid<bl)bl=mid;bc=mid;}
    }

    bool expansion() const { return vol_ratio>2.0 && fabs(ewm_drift)>2.0; }
};

struct Config {
    double atr_min, drift_min, vol_min;
    double sl_atr_mult;   // SL = entry +/- atr * mult
    double trail_atr_mult; // trail at atr * mult behind MFE
    int    max_hold_s;    // force close after N seconds
    bool   session_gate;  // block 17-22 UTC
    const char* label;
};

struct Stats {
    int n=0, wins=0;
    double net=0, peak=0, dd=0;
    double monthly[32]={};
    static constexpr int64_t MO_BASE=1695772800LL;

    void add(double pnl, int64_t ts_ms) {
        n++; net+=pnl; if(pnl>0)wins++;
        if(net>peak)peak=net;
        double d=peak-net; if(d>dd)dd=d;
        int mo=(int)((ts_ms/1000-MO_BASE)/2592000);
        if(mo>=0&&mo<32)monthly[mo]+=pnl;
    }
    void print(const char* lbl) const {
        if(!n){printf("  %-40s no trades\n",lbl);return;}
        int pos=0,tot=0; double ms=0;
        for(int i=0;i<32;i++)if(monthly[i]!=0){
            tot++; if(monthly[i]>0)pos++; ms+=monthly[i];
        }
        printf("  %-40s N=%4d WR=%5.1f%% Net=%+9.2f Avg=%+6.2f DD=%8.2f Mo=%d/%d\n",
               lbl,n,100.0*wins/n,net,net/n,dd,pos,tot);
        double cum=0;
        for(int i=0;i<32;i++){
            if(!monthly[i])continue; cum+=monthly[i];
            int64_t sec=MO_BASE+(int64_t)i*2592000;
            int yr=(int)(sec/31557600+1970),mo=(int)((sec%31557600)/2592000+1);
            char bar[41]={}; int b=std::min(40,(int)(fabs(monthly[i])/100));
            memset(bar,monthly[i]>0?'#':'.',b);
            printf("    %04d-%02d %+9.2f cum=%+9.2f %s\n",yr,mo,monthly[i],cum,bar);
        }
    }
};

int main(int argc, char** argv) {
    if(argc<2){printf("Usage: mce_sweep <ticks.csv>\n");return 1;}
    printf("Loading %s...\n",argv[1]);
    auto ticks=load(argv[1]);
    const size_t N=ticks.size();
    printf("Loaded %zu ticks\n\n",N);

    const double SPREAD=0.25, SLIP=0.05, COST=SPREAD+SLIP*2;
    const double RISK=50.0, USD_PT=100.0;
    const int64_t COOLDOWN_MS=120000LL; // 2min cooldown

    Config configs[]={
        // Original proven config
        {6.0, 5.0, 2.5, 1.5, 1.0, 3600, false, "ORIGINAL atr6 d5 v2.5"},
        // Slightly lower
        {5.0, 4.5, 2.0, 1.5, 1.0, 3600, false, "atr5.0 d4.5 v2.0"},
        {5.0, 4.0, 2.0, 1.5, 1.0, 3600, false, "atr5.0 d4.0 v2.0"},
        {4.5, 4.0, 2.0, 1.5, 1.0, 3600, false, "atr4.5 d4.0 v2.0"},
        {4.5, 3.5, 2.0, 1.5, 1.0, 3600, false, "atr4.5 d3.5 v2.0"},
        {4.0, 4.0, 2.0, 1.5, 1.0, 3600, false, "atr4.0 d4.0 v2.0"},
        {4.0, 3.5, 2.0, 1.5, 1.0, 3600, false, "atr4.0 d3.5 v2.0"},
        {4.0, 3.0, 2.0, 1.5, 1.0, 3600, false, "atr4.0 d3.0 v2.0"},
        {4.0, 3.0, 1.8, 1.5, 1.0, 3600, false, "atr4.0 d3.0 v1.8"},
        {3.5, 3.0, 1.8, 1.5, 1.0, 3600, false, "atr3.5 d3.0 v1.8"},
        // Session gated (block 17-22 UTC)
        {5.0, 4.0, 2.0, 1.5, 1.0, 3600, true,  "atr5.0 d4.0 v2.0 SESS"},
        {4.5, 3.5, 2.0, 1.5, 1.0, 3600, true,  "atr4.5 d3.5 v2.0 SESS"},
        {4.0, 3.0, 1.8, 1.5, 1.0, 3600, true,  "atr4.0 d3.0 v1.8 SESS"},
        // Wider SL -- survive volatility
        {5.0, 4.0, 2.0, 2.0, 1.5, 3600, false, "atr5.0 d4.0 sl2x trail1.5x"},
        {4.5, 3.5, 2.0, 2.0, 1.5, 3600, false, "atr4.5 d3.5 sl2x trail1.5x"},
        {4.0, 3.0, 2.0, 2.0, 1.5, 3600, false, "atr4.0 d3.0 sl2x trail1.5x"},
        // Tighter SL -- higher WR but more SL hits
        {5.0, 4.0, 2.0, 1.0, 0.8, 3600, false, "atr5.0 d4.0 sl1x trail0.8x"},
        {4.5, 3.5, 2.0, 1.0, 0.8, 3600, false, "atr4.5 d3.5 sl1x trail0.8x"},
    };
    const int NCFG=sizeof(configs)/sizeof(configs[0]);

    struct Pos {
        bool active=false, is_long=false;
        double entry=0, trail_sl=0, hard_sl=0, mfe=0, size=0;
        int64_t cooldown_until=0;
    };
    std::vector<Pos> pos(NCFG);
    std::vector<Stats> stats(NCFG);
    Indicators ind;

    for(size_t i=0;i<N;i++){
        const Tick& tk=ticks[i];
        const double mid=(tk.bid+tk.ask)*0.5;
        ind.update(tk.bid,tk.ask,tk.ts_ms);
        if(ind.atr<0.5) continue; // warmup

        const int hour=(int)((tk.ts_ms/1000)%86400/3600);

        for(int c=0;c<NCFG;c++){
            const Config& cfg=configs[c];
            Pos& p=pos[c];

            // Manage open position
            if(p.active){
                const double m=p.is_long?(tk.bid-p.entry):(p.entry-tk.ask);
                if(m>p.mfe){
                    p.mfe=m;
                    if(p.is_long) p.trail_sl=std::max(p.trail_sl, tk.bid-cfg.trail_atr_mult*ind.atr);
                    else          p.trail_sl=std::min(p.trail_sl, tk.ask+cfg.trail_atr_mult*ind.atr);
                }
                const bool max_hold=(tk.ts_ms-p.cooldown_until+COOLDOWN_MS)>(int64_t)cfg.max_hold_s*1000;
                const bool trail_hit=(p.is_long&&tk.bid<=p.trail_sl)||(!p.is_long&&tk.ask>=p.trail_sl);
                const bool hard_hit =(p.is_long&&tk.bid<=p.hard_sl )||(!p.is_long&&tk.ask>=p.hard_sl );
                if(trail_hit||hard_hit||max_hold){
                    double ep=hard_hit?p.hard_sl:(trail_hit?p.trail_sl:mid);
                    double pnl_pts=p.is_long?(ep-p.entry):(p.entry-ep);
                    double pnl_usd=pnl_pts*p.size*USD_PT-COST*p.size*USD_PT;
                    stats[c].add(pnl_usd,tk.ts_ms);
                    p.active=false;
                    p.cooldown_until=tk.ts_ms+COOLDOWN_MS;
                }
                continue;
            }

            // Entry gates
            if(tk.ts_ms<p.cooldown_until) continue;
            if(cfg.session_gate && hour>=17 && hour<22) continue;
            if(ind.atr < cfg.atr_min)    continue;
            if(fabs(ind.ewm_drift) < cfg.drift_min) continue;
            if(ind.vol_ratio < cfg.vol_min) continue;
            if(!ind.expansion()) continue;

            // Enter in direction of drift
            const bool is_long=ind.ewm_drift>0;
            const double ep=is_long?tk.ask:tk.bid;
            const double sl_pts=cfg.sl_atr_mult*ind.atr;
            const double hard_sl=is_long?ep-sl_pts:ep+sl_pts;
            const double trail_sl=hard_sl; // starts at hard SL
            const double sz=std::max(0.01,std::min(0.20,RISK/(sl_pts*USD_PT)));

            p.active=true; p.is_long=is_long; p.entry=ep;
            p.trail_sl=trail_sl; p.hard_sl=hard_sl;
            p.mfe=0; p.size=sz;
            // store entry time in cooldown_until temporarily
            p.cooldown_until=tk.ts_ms; // will be overwritten on close
        }
    }

    // Print results sorted by net P&L
    std::vector<int> order(NCFG);
    for(int i=0;i<NCFG;i++)order[i]=i;
    std::sort(order.begin(),order.end(),[&](int a,int b){return stats[a].net>stats[b].net;});

    printf("%-40s %4s %5s %9s %6s %8s %s\n","Config","N","WR%","Net","Avg","DD","Mo+/tot");
    printf("%s\n",std::string(85,'-').c_str());
    for(int idx:order) stats[idx].print(configs[idx].label);

    return 0;
}

File: backtest/mce_sweep.py
#!/usr/bin/env python3
"""
MCE threshold sweep -- find the optimal ATR/drift/vol thresholds
that maximize profitable trades while preserving the 69% WR signal quality.

The only profitable engine in 2yr backtest is MCE at original thresholds.
We need more trades. This sweeps thresholds to find where edge degrades.

Key question: at ATR=4.5, drift=4, vol=2.0 -- does WR drop below 50%?
"""

import sys, math, collections, csv
from dataclasses import dataclass

DATA = sys.argv[1] if len(sys.argv) > 1 else "/Users/jo/Tick/2yr_XAUUSD_tick.csv"
SPREAD = 0.25
RISK   = 30.0
USD_PT = 100.0

def make_ewm(hl):
    s = {"v": None, "ts": None}
    def u(v, ts):
        if s["v"] is None: s["v"]=v; s["ts"]=ts; return v
        dt=max(ts-s["ts"],0.001); a=1-math.exp(-dt*0.693147/hl)
        s["v"]=a*v+(1-a)*s["v"]; s["ts"]=ts; return s["v"]
    return u

@dataclass
class Pos:
    active:bool=False; is_long:bool=True
    entry:float=0; sl:float=0; tp:float=0
    size:float=0; entry_ts:int=0
    mfe:float=0; mae:float=0

class Res:
    def __init__(self,label): self.label=label; self.trades=[]; self.monthly=collections.defaultdict(float)
    def add(self, pnl, ts, is_win):
        self.trades.append((pnl,ts,is_win))
        mo = f"{1970+ts//31557600:04d}-{(ts%31557600)//2592000+1:02d}"
        self.monthly[mo] += pnl
    def summary(self):
        if not self.trades: return f"{self.label:<35} {'no trades':>10}"
        pnls=[t[0] for t in self.trades]
        wins=[p for p in pnls if p>0]
        net=sum(pnls); wr=len(wins)/len(pnls)*100
        peak=eq=dd=0
        for p in pnls:
            eq+=p; peak=max(peak,eq); dd=max(dd,peak-eq)
        pos_mo=sum(1 for v in self.monthly.values() if v>0)
        return (f"{self.label:<40} n={len(pnls):>4} WR={wr:>5.1f}% "
                f"net=${net:>+9.2f} avg=${net/len(pnls):>+6.2f} "
                f"dd=${dd:>7.2f} pos_mo={pos_mo}/{len(self.monthly)}")

# Threshold combinations to test
# (atr_min, drift_min, vol_min, sl_mult, label)
configs = [
    # Original -- proven 69% WR
    (6.0, 5.0, 2.5, 1.5, "ORIGINAL atr6 d5 v2.5"),
    # Slightly lower
    (5.0, 4.5, 2.0, 1.5, "atr5 d4.5 v2.0"),
    (5.0, 4.0, 2.0, 1.5, "atr5 d4.0 v2.0"),
    (4.5, 4.0, 2.0, 1.5, "atr4.5 d4.0 v2.0"),
    (4.5, 3.5, 2.0, 1.5, "atr4.5 d3.5 v2.0"),
    (4.0, 4.0, 2.0, 1.5, "atr4.0 d4.0 v2.0"),
    (4.0, 3.5, 2.0, 1.5, "atr4.0 d3.5 v2.0"),
    (4.0, 3.0, 2.0, 1.5, "atr4.0 d3.0 v2.0"),
    (4.0, 3.0, 1.8, 1.5, "atr4.0 d3.0 v1.8"),
    (3.5, 3.0, 1.8, 1.5, "atr3.5 d3.0 v1.8"),
    # With session gate (Asia+London+early NY: 22-17 UTC, block 17-22)
    (5.0, 4.0, 2.0, 1.5, "atr5 d4.0 v2.0 SESSION"),
    (4.5, 3.5, 2.0, 1.5, "atr4.5 d3.5 v2.0 SESSION"),
    (4.0, 3.0, 1.8, 1.5, "atr4.0 d3.0 v1.8 SESSION"),
]

results = {cfg[4]: Res(cfg[4]) for cfg in configs}
positions = {cfg[4]: Pos() for cfg in configs}
last_close = {cfg[4]: 0 for cfg in configs}

ewm30 = make_ewm(30)
ewm120 = make_ewm(120)
atr = 0.0
vs = vl = 0.0  # vol ratio EWMs

# PDH/PDL
pdh = pdl = 0.0
today_hi = 0.0; today_lo = 1e9; cur_day = -1

last_ts = 0; n = 0
print(f"Loading {DATA}...")

with open(DATA) as f:
    for line in f:
        n += 1
        if n % 20_000_000 == 0: print(f"  {n//1_000_000}M rows...")
        line = line.strip()
        if not line: continue
        p = line.split(',')
        if len(p) < 4: continue
        try:
            bid=float(p[2]); ask=float(p[3])
        except: continue
        if bid<=0 or ask<=bid: continue
        try:
            ds=p[0]; ts_s=p[1]
            y=int(ds[:4]); mo=int(ds[4:6]); dy=int(ds[6:8])
            h=int(ts_s[:2]); mi=int(ts_s[3:5]); se=int(ts_s[6:8])
            if mo<=2: y-=1; mo+=12
            days=365*y+y//4-y//100+y//400+(153*mo+8)//5+dy-719469
            ts=(days*86400+h*3600+mi*60+se)
        except: continue

        mid=(bid+ask)*0.5
        if last_ts>0 and ts-last_ts>3600:
            ewm30=make_ewm(30); ewm120=make_ewm(120); vs=vl=0

        last_ts=ts
        dt=1.0  # approximate

        em30=ewm30(mid,ts); em120=ewm120(mid,ts)
        drift=mid-em30

        # ATR (tick EWM scaled to bar-equivalent)
        spread_now=ask-bid
        if atr==0: atr=spread_now*20
        else: atr=0.002*spread_now*20+0.998*atr

        # Vol ratio: short/long EWM of |drift|
        ab=abs(drift)
        vs=0.1*ab+0.9*vs; vl=0.01*ab+0.99*vl
        vol_ratio=vs/vl if vl>1e-9 else 1.0

        hour=(ts%86400)//3600

        # PDH/PDL
        day=ts//86400
        if day!=cur_day:
            if cur_day>=0: pdh=today_hi; pdl=(today_lo if today_lo<1e8 else 0)
            today_hi=mid; today_lo=mid; cur_day=day
        else:
            if mid>today_hi: today_hi=mid
            if mid<today_lo: today_lo=mid

        for cfg in configs:
            atr_min,drift_min,vol_min,sl_mult,label = cfg
            session_gate = "SESSION" in label
            if session_gate and (17<=hour<=21): continue  # block late NY

            pos=positions[label]; res=results[label]

            if pos.active:
                m=(mid-pos.entry) if pos.is_long else (pos.entry-mid)
                pos.mfe=max(pos.mfe,m); pos.mae=min(pos.mae,m)
                sl_hit=(pos.is_long and bid<=pos.sl) or (not pos.is_long and ask>=pos.sl)
                tp_hit=(pos.is_long and ask>=pos.tp) or (not pos.is_long and bid<=pos.tp)
                time_exit=(ts-pos.entry_ts)>=600
                if sl_hit or tp_hit or time_exit:
                    ep=pos.sl if sl_hit else (pos.tp if tp_hit else (bid if pos.is_long else ask))
                    pnl_pts=(ep-pos.entry) if pos.is_long else (pos.entry-ep)
                    pnl_usd=pnl_pts*pos.size*USD_PT - SPREAD*pos.size*USD_PT
                    res.add(pnl_usd, ts, pnl_usd>0)
                    pos.active=False; last_close[label]=ts
                continue

            if ts-last_close[label]<120: continue
            if atr<atr_min: continue
            if vol_ratio<vol_min: continue
            if abs(drift)<drift_min: continue

            # Entry: follow the drift direction (MCE logic -- ride the impulse)
            is_long=drift>0
            sl_pts=atr*sl_mult*0.05  # scale tick ATR to tradeable SL
            sl_pts=max(sl_pts, 1.5)  # minimum 1.5pt SL
            ep=ask if is_long else bid
            sl_px=ep-sl_pts if is_long else ep+sl_pts
            tp_px=ep+sl_pts*2 if is_long else ep-sl_pts*2
            sz=max(0.01,min(0.20, RISK/(sl_pts*USD_PT)))
            sz=math.floor(sz/0.001)*0.001

            positions[label]=Pos(True,is_long,ep,sl_px,tp_px,sz,ts,0,0)

print(f"\nLoaded {n:,} rows\n")
print(f"{'Config':<40} {'N':>5} {'WR%':>6} {'Net':>10} {'Avg':>7} {'DD':>8} {'POS/MO':>7}")
print("─"*85)
for cfg in configs:
    print(results[cfg[4]].summary())

# Detailed monthly for configs that show positive months
print("\n\nDETAIL -- configs with any profitable months:")
for cfg in configs:
    r=results[cfg[4]]
    if not r.trades: continue
    pos_mo=sum(1 for v in r.monthly.values() if v>0)
    if pos_mo < 3: continue
    print(f"\n{cfg[4]}:")
    cum=0
    for mo,v in sorted(r.monthly.items()):
        cum+=v
        bar='#'*int(v/30) if v>0 else '.'*int(abs(v)/30)
        print(f"  {mo}  {v:>+8.2f}  cum={cum:>+9.2f}  {bar[:40]}")

